ath79: rework ar9342 based ubnt devices

PolynomialDivision opened the pull request:
Please notice:
You have to apply https://github.com/openwrt/openwrt/pull/3536#pullrequestreview-516355511

... Testing all devices again ... (Draft)

- [x] Nanobeam AC Gen2
- [x] Litebeam AC Gen2
- [x] Nanostation AC Loco
- [x] Lite AP 120
- [x] Nanobeam AC
- [ ] Nanostation AC
- [x] Powerbeam AC Gen2

Images for missing devices:
[nanobeam.zip](https://github.com/openwrt/openwrt/files/5435569/nanobeam.zip)
[nanostation.zip](https://github.com/openwrt/openwrt/files/5435574/nanostation.zip)
[powerbeam.zip](https://github.com/openwrt/openwrt/files/5435576/powerbeam.zip)
[lap-120.zip](https://github.com/openwrt/openwrt/files/5435577/lap-120.zip)

PolynomialDivision committed:
ath79: merge ar9342 dts files

Most ubnt-wa devices differ only in the number of ethernet ports.
Therefore we introduce another 1port, or 2port definition.

Signed-off-by: Nick Hainke <vincent@systemli.org>

adschm commented:
Please "fix" the ac loco first in a separate patch, and then do the DTSI creation in a separate patch (which will be cosmetic then, i.e. not change functionality).

Litebeam gen1 can also be dealt with in the DTSI patch.

What about the remaining devices, nanobeam ac and powerbeam-5ac-gen2? If you create general DTSI files with titles like 1port and 2port, they should actually apply to all one-port and two-port devices and not just a fraction.

adschm commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -194,6 +194,14 @@ define Device/ubnt_nanobeam-ac
+define Device/ubnt_nanobeam-ac-gen2-wa
+  $(Device/ubnt-wa)
+  DEVICE_MODEL := NanoBeam AC
+  DEVICE_VARIANT := Gen2
```
If we have a variant here, it would be helpful to also add a device variant to "gen1" ...

PolynomialDivision replied:
Should I rename `ubnt_nanobeam-ac-gen1-wa`? This will "break" sysupgrade for users since they have to force. :/

adschm replied:
I was mainly referring to DEVICE_VARIANT variable for now, as that one is just informational and won't break anything.

I would rather not change compatible/image name at this point. Is the wa suffix needed for this device? Since we don't have it for the other ar9342 ubnt wa devices so far, I'd rather stay consistent to those unless there is a non-wa nanobeam ac...

adschm replied:
BTW the sysupgrade couldn't be healed by simply adding the old name to supported devices:

SUPPORTED_DEVICES += ubnt,nanobeam-ac

PolynomialDivision committed:
ath79: Add support for Ubiquiti NanoBeam AC Gen2

CPU:         Atheros AR9342 rev 3 SoC
RAM:         64 MB DDR2
Flash:       16 MB NOR SPI
WLAN 2.4GHz: Atheros AR9342 v3 (ath9k)
WLAN 5.0GHz: QCA988X
Ports:       2x GbE

Flashing procedure is identical to other ubnt devices.
https://openwrt.org/toh/ubiquiti/common

Flashing through factory firmware
1. Ensure firmware version v8.7.0 is installed.
   Up/downgrade to this exact version.
2. Patch fwupdate.real binary using
   `hexdump -Cv /bin/ubntbox | sed 's/14 40 fe 27/00 00 00 00/g' | \
    hexdump -R > /tmp/fwupdate.real`
3. Make the patched fwupdate.real binary executable using
   `chmod +x /tmp/fwupdate.real`
4. Copy the squashfs factory image to /tmp on the device
5. Flash OpenWrt using `/tmp/fwupdate.real -m <squashfs-factory image>`
6. Wait for the device to reboot
(copied from Ubiquiti NanoBeam AC and modified)

To keep it consistent, we will add the gen1 variant to
the nanobeam ac gen1.

Signed-off-by: Nick Hainke <vincent@systemli.org>

PolynomialDivision commented:
> What about the remaining devices, nanobeam ac and powerbeam-5ac-gen2? If you create general DTSI files with titles like 1port and 2port, they should actually apply to all one-port and two-port devices and not just a fraction.

I really would love to add them, but I can not test those changes. :S I only have 3 of those devices.

PolynomialDivision commented:
@adschm I believe the nanobeam ac will break with 5.4 kernel, too. Since it has the same wrong ethernet phy description like the nanostation ac loco. I will add a further "fix" commit for that, but I can not test if that is actually fixing the issue...?

PolynomialDivision committed:
ath79: fix nanobeam ac ethernet interface

In 4.14 the delays were not cleared, so setting "rgmii" as phy-mode
did not affect delays set by the bootloader. With 5.4 kernel the
situation changed and the ethernet interface stopped working.

"rgmii" requires rx and tx delays depending on the hardware circuit
and wiring. The mac or the phy can add these delays.
- "rgmii":  delays are controlled by the mac
- "rgmii-id": delays are controlled by the phy
More Information in Linux Kernel Tree:
Documentation/devicetree/bindings/net/ethernet-controller.yaml

"rgmii" should be the preferred mode, which allows the mac layer to
turn off the dealys completely if they are not needed. However, the
delays are not set correctly, which causes the ethernet interface
to be broken. Just taking the ethernetpart from the litebeam ac gen2
will fix the issue.

Explained-by: David Bauer <mail@david-bauer.net>
Signed-off-by: Nick Hainke <vincent@systemli.org>

adschm commented:
Just look who added the other devices to the repo and ping them here ...

adschm commented:
> I really would love to add them, but I can not test those changes. :S I only have 3 of those devices.

So, this means you _think_ there actually are _only two_ ethernet setup variants for all ar9342 ubnt wa devices?

PolynomialDivision commented:
> So, this means you _think_ there actually are _only two_ ethernet setup variants for all ar9342 ubnt wa devices?

Yes I only think that they have at maximum 2 ethernet ports. All the devices I looked up have 1 or 2 ethernet ports. I searched for a datasheet for the ar9342 soc, but for now I did not find a datasheet. :/

adschm commented:
Just merged a few of the easy commits.

PolynomialDivision commented:
> Just merged a few of the easy commits.

Whups. Actually I wanted to ask @blocktrron before merge if it is okay, like I wrote in the commit message that he explained me why the ethernet stopped working, to give him credit. I hope this is okay. :S

PolynomialDivision commented:
@blocktrron is fine with everything! :+1: Thanks for merging. Should I rebase?

adschm commented:
Yes, please rebase. And I would add the "Gen1" in the same commit where you add Gen2, I don't think a separate commit is necessary here (as it contains no "functional" change after all, and is only necessary due to adding gen2).

PolynomialDivision commented:
Is GitHub broken? o.O It tells me I have conflicts, but there are no conflicts? In addition, I removed the `-wa` extension for the nanobeam ac gen2 everywhere. I'm compiling... Deleted the toolchain so it takes a while. :/

PolynomialDivision commented:
I tested all devices I have. I zipped all missing images:
https://file.io/BMmuynD5kcYD

neheb commented:
I've seen the GitHub issue as well. Just ignore it.

karlp commented on target/linux/ath79/dts/ar9342_ubnt_nanobeam-ac.dts:
```diff
@@ -51,11 +50,11 @@
 	status = "okay";
 
 	/* default for ar934x, except for 1000M and 10M */
-	pll-data = <0x06000000 0x00000101 0x00001313>;
```
So, is the comment wrong now too?  or was it wrong before?

PolynomialDivision replied:
Not sure, I used the values from the  Litebeam AC Gen2, since it is a 1port devices that has a working ethernet switch. The next commit merges all the 1port devices together.

karlp commented on target/linux/ath79/dts/ar9342_ubnt_nanobeam-ac.dts:
```diff
@@ -51,11 +50,11 @@
 	mtd-mac-address = <&art 0x0>;
 
-	phy-mode = "rgmii";
+	phy-mode = "rgmii-id";
```
the difference here is probably more worthy of commenting...

PolynomialDivision replied:
"In 4.14 the delays were not cleared, so setting "rgmii" as phy-mode did not affect delays set by the bootloader. With 5.4 kernel the situation changed and the ethernet interface stopped working." ;)
That is in the commit message.

karlp replied:
.... yes.... I can read that too.  How about...

```
In 5.4, phy timings were automatically added by the mac, switch to using the new rgmii-id which uses the old behaviour
```

Could this be worked around by specifying the actual delays instead of just turning off the autosetting?  I don't personally find, "oh, in 5.4 it stopped working, so we changed some magic" to be very descriptive.

PolynomialDivision replied:
Okay, I will write a new commit message. ;)
How can I get the rx and tx delays? Just by try and error? Any data sheet, somwhere?

PolynomialDivision replied:
```


In 4.14 the delays were not cleared, so setting "rgmii" as phy-mode
did not affect delays set by the bootloader. With 5.4 kernel the
situation changed and the ethernet interface stopped working.

"rgmii" requires rx and tx delays depending on the hardware circuit
and wiring. The mac or the phy can add these delays.
- "rgmii":  delays are controlled by the mac
- "rgmii-id": delays are controlled by the phy
More Information in Linux Kernel Tree:
Documentation/devicetree/bindings/net/ethernet-controller.yaml

"rgmii" should be the preferred mode, which allows the mac layer to
turn off the dealys completely if they are not needed. However, the
delays are not set correctly, which causes the ethernet interface
to be broken. Just taking the ethernetpart from the litebeam ac gen2
will fix the issue.
```

I will try to findout working delays and get rgmii running.

PolynomialDivision replied:
@karlp is the commit message okay?

PolynomialDivision commented:
**rebased on master**

PolynomialDivision commented:
I know tested 5 out of 7 devices. Roger does not have the LAP-120, anymore but tested the Powerbeam. :/ Tobias is not responding.
I would guess that everything works? I would like to have the nanobeam ac gen2 merged. :)

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -358,6 +358,7 @@ ubnt,rocket-m)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "green:link4" "wlan0" "76" "100"
 	;;
 ubnt,nanobeam-ac|\
+ubnt,nanobeam-ac-gen2|\
```
No rssileds defined in DTS?

adschm commented on target/linux/ath79/dts/ar9342_ubnt_nanobeam-ac-gen2.dts:
```diff
@@ -0,0 +1,14 @@
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+};
```
This is already set in ar9342_ubnt_wa ...

adschm commented on target/linux/ath79/dts/ar9342_ubnt_nanobeam-ac-gen2.dts:
```diff
@@ -0,0 +1,14 @@
+
+/ {
+	compatible = "ubnt,nanobeam-ac-gen2", "ubnt,wa", "qca,ar9342";
+	model = "Ubiquiti NanoBeam AC Gen2";
```
Add a (WA) after Gen2 ...

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -59,6 +59,7 @@ ath79_setup_interfaces()
 	ubnt,lap-120|\
 	ubnt,litebeam-ac-gen2|\
 	ubnt,nanobeam-ac|\
+	ubnt,nanobeam-ac-gen2|\
```
I thought this has two ports?

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -582,6 +583,7 @@ ath79_setup_macs()
 		label_mac=$wan_mac
 		;;
 	ubnt,litebeam-ac-gen2|\
+	ubnt,nanobeam-ac-gen2|\
```
Did you verify this?

PolynomialDivision replied:
Yes. I tried the call `mtd_get_mac_binary art 0x5006` that gave me some useful stuff back.

I tried again `caldata_extract "art" 0x5000 0x844` and it was crashing my nanobeam ac gen2?

adschm replied:
art 0x5006 shouldn't just be "useful", but matching the MAC address on the label ...?

PolynomialDivision replied:
I mean it is beginning with `E0:63:DA` and that is ubiquiti networks. ;) I can not easily access the antenna, but maybe I can flash airos again and check mac address, then.

adschm replied:
flashing airos won't help you in order to find out what's printed on the label. If you are not sure what's on the label, don't set up a label mac address, and just remove that part.

PolynomialDivision replied:
*removed* ;)

PolynomialDivision replied:
added again, since I could verify that is matches with the label mac on the box in wihch it was delivered

adschm commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -190,10 +190,19 @@ TARGET_DEVICES += ubnt_litebeam-ac-gen2
 define Device/ubnt_nanobeam-ac
   $(Device/ubnt-wa)
   DEVICE_MODEL := NanoBeam AC
+  DEVICE_VARIANT := Gen1
```
You should also do the same change to the Gen1 DTS model property, i.e. add "Gen1" there.

adschm commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -190,10 +190,19 @@ TARGET_DEVICES += ubnt_litebeam-ac-gen2
+  $(Device/ubnt-wa)
+  DEVICE_MODEL := NanoBeam AC
+  DEVICE_VARIANT := Gen2
+  DEVICE_PACKAGES += kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct rssileds
```
As stated above: Do you actually have rssileds?

adschm replied:
Just for the record: Why does nobody ever answer a question in this place?

So, now, I have to go through all these closed conversations, open all of them manually, and find out what the answer to my question could be by comparing code.

PolynomialDivision replied:
Yes. It has rssi leds. I have added the code that defines the rssileds from other ar9342 devices. @ae6xe Could u test the changes? Otherwise I have to climb on a roof again.

ae6xe replied:
@PolynomialDivision , yes will test today on NanoBeam AC gen2 and Nanostation AC.

@adschm re: "Do you actually have rssileds".   Both of these devices have blue leds for 1) Power; 2) main; 3) secondary/LAN2; 4) multiple LEDs for RSSI.

ae6xe replied:
@PolynomialDivision this commit has conflicts to merge with master. Rebase pull with master, so I can build, then test?

rogerpueyo commented:
@PolynomialDivision ,

I could test the patch on the LiteAP ac 120 for a couple of days It worked well, 10/100/1000.

PolynomialDivision commented:
> @PolynomialDivision ,
> 
> I could test the patch on the LiteAP ac 120 for a couple of days It worked well, 10/100/1000.

Thanks for testing! :)

ae6xe commented:
I have a nanobeam AC gen2 and an nanostation AC newly purchased.  (I will have more AC devices arriving soon.)   If I can assist in testing, or otherwise, let me known.

Joe AE6XE

PolynomialDivision commented:
> I have a nanobeam AC gen2 and an nanostation AC newly purchased. (I will have more AC devices arriving soon.) If I can assist in testing, or otherwise, let me known.
> 
> Joe AE6XE

Thanks! :)
You can use the current openwrt master and test on nanostation ac, since all commits are already included that affect those device. :)
The last commit is the Nanobeam AC Gen2.

@adschm I found the box in which my Nanobeam AC Gen2 was delivered and found the MAC. It matches with the mac stored in art. So I added that again.

adschm commented:
Unfortunately, this still lacks the proper setup in 02_network.

PolynomialDivision commented:
> _Please just report whether it's the same block as for nanostation-ac, so I can add it locally instead of having to diff the commit another time._

Sry, what do u mean with that? Sry for closing everything. This is the first time I'm adding device support. Next time I won't close them.

adschm commented:
@ae6xe I've pushed a rebased version here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/nanobeam

I've reverted two netifd commits there because there have been vlan problems reported with latest master.

adschm commented:
Just learned that the second netifd commit fixed the first one, so the reverts can be removed again (or just kept), both should work for testing.

PolynomialDivision commented:
I activated the same scrips for the nanobeam ac gen2 as the nanostation ac, except the GPIO stuff:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/03_gpio_switches#L79-L81

@ae6xe Is there some poe switch on the nanobeam ac gen2?

PolynomialDivision commented:
I don't understand what poe switch should be on the nanostation ac, I looked into the documentation and there is no switch

https://dl.ubnt.com/qsg/NS-5AC/NS-5AC_DE.html

rogerpueyo commented:
> there is no switch

_10/100/1000-Ethernet-Port für Bridging mit Unterstützung von softwarekonfigurierbarem PoE-Passthrough_:I'd say the PoE passthrough is software controlled, by setting a certain GPIO on or off. So there's no physical switch.

adschm commented:
From 03_gpio_switches:

```
ubnt,nanostation-ac)
	ucidef_add_gpio_switch "poe_passthrough" "PoE Passthrough" "3"
	;;
```

ae6xe commented:
tested https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/nanobeam on a Nanostation 5GHz AC.   All led lights are working as expected -- full signal strength of 4 blue LEDs lite when connecting a client.   Both LAN and WAN ports are working as expected.

[board.json.txt](https://github.com/openwrt/openwrt/files/5539615/board.json.txt)

ae6xe commented:
I can confirm on the Nanostation 5GHz AC, gpio3 indeed turns POE passthrough on/off.

PolynomialDivision commented:
> This still lacks port setup in 02_network. Currently, you set up two separate interfaces for lan and wan there (eth0/eth1 via the default case), but I think these are two ports connected to a single switch?

Yes.

> I can confirm on the Nanostation 5GHz AC, gpio3 indeed turns POE passthrough on/off.

So I will add the nanobeam ac gen2, to the nanostation ac gpio switxh directoy. Thanks for testing.

adschm commented:
> > This still lacks port setup in 02_network. Currently, you set up two separate interfaces for lan and wan there (eth0/eth1 via the default case), but I think these are two ports connected to a single switch?
> 
> Yes.
> 
> > I can confirm on the Nanostation 5GHz AC, gpio3 indeed turns POE passthrough on/off.
> 
> So I will add the nanobeam ac gen2, to the nanostation ac gpio switxh directoy. Thanks for testing.

And please then report back with a final test of everything on the nanobeam gen2 ...

ae6xe commented:
>>So I will add the nanobeam ac gen2, to the nanostation ac gpio switxh directoy. Thanks for testing.
>And please then report back with a final test of everything on the nanobeam gen2 ...

I'm not entirely sure what ready-to-build reference point to test with, please confirm after "add [gpio passthough to] the nanobeam ac gen2" , where to obtain.

PolynomialDivision commented:
> > > So I will add the nanobeam ac gen2, to the nanostation ac gpio switxh directoy. Thanks for testing.
> > > And please then report back with a final test of everything on the nanobeam gen2 ...
> 
> I'm not entirely sure what ready-to-build reference point to test with, please confirm after "add [gpio passthough to] the nanobeam ac gen2" , where to obtain.

Check if gpio3 is working
```
echo 3 > /sys/class/gpio/export
echo out > /sys/class/gpio/gpio3/direction
echo 1 > /sys/class/gpio/gpio3/value
```

PolynomialDivision commented:
Or even easier, activate it under `/etc/config/system` xD

ae6xe commented:
For completeness, this is how I previously tested POE pass through on the Nanostation 5GHz AC:

echo 3 > /sys/class/gpio/export
echo out > /sys/class/gpio/gpio3/direction
echo 1 > /sys/class/gpio/gpio3/value

This test also succeeds on the Nanobeam 5GHz AC gen2.  On both devices an older Ubiquiti AirCam is powered on/off.

PolynomialDivision commented:
Whups. :/
I did not push the gpio uci commit. I pushed it now, and I'm compiling and testing again.

pepe2k commented:
@PolynomialDivision I change title of your PR to be less generic, hope you don't mind.

Cheers,
Piotr

PolynomialDivision commented:
@pepe2k I don't mind ;) It is only the nanobeam ac gen2 left.

I accidentily flashed the litebeam ac gen2 image, that is why it tool some time to flash the nanobeam ac gen2 again. With the last changes I have now 

```ùci
onfig gpio_switch 'poe_passthrough'
	option name 'PoE Passthrough'
	option gpio_pin '3'
	option value '0'
```
in `/etc/config/system`. :)
@adschm  I don't see any issues with this nanobeam ac gen2, anymore.

PolynomialDivision commented:
But it is now identical to nanostation ac. Is there a way to just build nanostation ac for nanobeam ac gen2? Otherwise, we have unnecessary code duplication?

adschm commented:
> But it is now identical to nanostation ac. Is there a way to just build nanostation ac for nanobeam ac gen2? Otherwise, we have unnecessary code duplication?

There is nothing really helpful for these cases, and having to use images with different names is typically misleading, particularly if the is a nanobeam ac (gen1) around that might be mistaken for gen2.

So, I prefer to have a separate device/image in OpenWrt, and reduce duplicate code with the available measures like DTSI.

@PolynomialDivision So, can the nanobeam ac gen2 now be considered "tested" with the most recent patch state?

adschm commented:
Anyway, merged, thanks.
